src/swsssdk/dbconnector.py

```diff
@@ -26,165 +29,384 @@ def load_sonic_db_config(sonic_db_file_path=SONIC_DB_CONFIG_FILE):
+        ns_instances = []
+        if namespace == 'all':
+            ns_instances = self.get_db_connection_instances()
+        else:
+            ns_instances.append(namespace)
+
+        total_clients = 0
+        for instance in ns_instances:
+            db_id = self.get_dbid(db_name, instance)
+            total_clients += self.db_intfs[instance].publish(db_id, channel, message)
+        return total_clients
+
+    def expire(self, db_name, key, timeout_sec, namespace='none'):
+        if namespace == 'none':
+            namespace = self.ns_name
 
-    def expire(self, db_name, key, timeout_sec):
-        db_id = self.get_dbid(db_name)
-        return super(SonicV2Connector, self).expire(db_id, key, timeout_sec)
+        ns_instances = []
+        if namespace == 'all':
+            ns_instances = self.get_db_connection_instances()
+        else:
+            ns_instances.append(namespace)
+
+        for instance in ns_instances:
+            db_id = self.get_dbid(db_name, instance)
+            self.db_intfs[instance].expire(db_id, key, timeout_sec)
 
-    def exists(self, db_name, key):
-        db_id = self.get_dbid(db_name)
-        return  super(SonicV2Connector, self).exists(db_id, key)
+    def exists(self, db_name, key, namespace='none'):
+        if namespace == 'none':
+            namespace = self.ns_name
 
-    def keys(self, db_name, pattern='*', *args, **kwargs):
-        db_id = self.get_dbid(db_name)
-        return super(SonicV2Connector, self).keys(db_id, pattern, *args, **kwargs)
+        ns_instances = []
+        if namespace == 'all':
+            ns_instances = self.get_db_connection_instances()
+        else:
+            ns_instances.append(namespace)
 
-    def get(self, db_name, _hash, key, *args, **kwargs):
-        db_id = self.get_dbid(db_name)
-        return super(SonicV2Connector, self).get(db_id, _hash, key, *args, **kwargs)
+        for instance in ns_instances:
+            db_id = self.get_dbid(db_name, instance)
 
-    def get_all(self, db_name, _hash, *args, **kwargs):
-        db_id = self.get_dbid(db_name)
-        return super(SonicV2Connector, self).get_all(db_id, _hash, *args, **kwargs)
+            # We find the key, return back success.
+            if self.db_intfs[instance].exists(db_id, key):
+                return 1
 
-    def set(self, db_name, _hash, key, val, *args, **kwargs):
-        db_id = self.get_dbid(db_name)
-        return super(SonicV2Connector, self).set(db_id, _hash, key, val, *args, **kwargs)
+        # We reach here, didn't find the key any where
+        return 0
 
-    def delete(self, db_name, key, *args, **kwargs):
-        db_id = self.get_dbid(db_name)
-        return super(SonicV2Connector, self).delete(db_id, key, *args, **kwargs)
+    def keys(self, db_name, pattern='*', namespace='none', *args, **kwargs):
+        if namespace == 'none':
+            namespace = self.ns_name
+
+        data = []
+        ns_instances = []
+        if namespace == 'all':
```

Comment:
>if namespace == 'all': [](start = 8, length = 22)

This 'all' namespace seems not fundamental feature of SonicV2Connector class. Could you extract this feature outside this class?

Reply:
Understand your comment that "all", "none" etc could be valid namespace names. 

I have made a few changes to the options, as below to connect to the databases. Introduced a new parameter "mode"

"namespace"
	'default' : it connects to the local 'db_name' instance.
	<asic0,asic1 ...asicn> : it connects to "db_name" instance in that namespace
"mode"
	'exclusive' : connect exclusively to the db instance in the namespace specified.
	'all' : connect to the db instance of the "db_name" in all namespaces.

I would like to have the SonicV2 Connector class and ConfigDBConnector class able to connect to a specific db_instance exclusively -- OR -- connect to the db_instances in all namespaces.

I suggest to keep it this way as we have seen many use cases where we need to get/set tables in a similar fashion on the db_instances in all namespaces. With the utility class supporting 'all' we can avoid adding for loops every where in the code ... eg: in get/get_all in case of SNMP, Config CLI start/shut all BGP neighbors etc

If user don't specifiy these parameters, ( namespace, mode ) will take ('default' , 'exclusive') which connects to the specific db_instance in the local redis server. This helps to retain the current behavior of these utility classes, if the user don't pass any arguments.

Reply:
I thought about the various use cases and came up with this new approach for user to pass a list of namespaces who he wants to connect to. I thought about having separate classes for single namespace and multi namespace -- but found that there is so much of code duplication if I do so.

Additionally removing the namespace parameter from the APIs helped reduce the new code added.